Fix Part of #4938: Profile Configuration and Migration

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fix Part of #4938:  Configure profile new creation and migration of existing profiles.

## When Onboarding V2 is Enabled:
### Sole Learner Onboarding
Once the learner has created a profile, and have landed on the introduction screen, the profile is marked as `started_profile_oboarding`, and a corresponding event log is added.
This configuration facilitates the user resuming onboarding, if they didn't complete it the first time round.

After selecting the audio language, the profile is logged in, and routed to the home screen.
On the home screen, the profile is updated to indicate `completed_profile_oboarding` and 3 events are expected for the first time login flow, in order:
1.  Open home event
2.  Profile Onboarding Completed event
3.  App Onboarding Completed event

The  App Onboarding Completed event is only logged for the first profile on the device, while the  Profile Onboarding Completed event is logged for every profile on first login.

### Supervisor Profile Onboarding
From the Profile Type screen, the supervisor flow launches the profile chooser screen, showing the Admin profile, which on click launches the homescreen.

The same 3 events as above are expected for the first time login flow.

### Login Routing
Returning users should be routed to an appropriate landing page as follows:
A sole learner profile will be routed directly to their home screen 

Returning admin profiles and non-solo learner profiles will always be routed to the profile selection screen

A sole learner who started, but did not finish onboarding would be routed to the Introduction Screen.

Returning sole and none-sole learners created when the feature flag was disabled, would be directed to the onboarding screen. Profile migration happens in place when profiles are fetched by the controller.  

### Profile Creation and Migration
For migration purposes, the `getProfile()` function has been updated to compute the `profileType` field when the feature flag is enabled. Existing Admin profiles will be migrated to have the SUPERVISOR type, existing Admin profiles with no pins set will be migrated to have the SOLE_LEARNER type while the remaining accounts will be of the ADDITIONAL_LEARNER type. New profiles created will also contain the respective enum type based on the intended categorization. 
Flag off then on:
[device-2024-06-21-073957.webm](https://github.com/oppia/oppia-android/assets/59600948/7584ce32-305b-4743-878e-64c94dc52e66)

Flag on then off:
[device-2024-06-21-074114.webm](https://github.com/oppia/oppia-android/assets/59600948/87860084-c75c-4b6c-a37f-c2459a3e7794)

### Misc
- There are general Kdoc formatting fixes all over the place.
- `ProfileTestHelper` Has been modified to create a sole, pinless admin profile.
- There are other minor changes to fix tests that have been impacted by changes in this PR.
- The changes in `DeprecationController` are purely visual to make the code more readable.
- `ClassroomListFragmentTest` required refactor to support toggling multiple feature flags.

### Profile Onboarding Events
|Onborading Flow v1| Onboarding Flow v2|
|---|---|
|<img src="https://github.com/oppia/oppia-android/assets/59600948/990c1c0a-ce71-4685-b986-bb4396ce28e6" width="350">|<img src="https://github.com/oppia/oppia-android/assets/59600948/c5eca79b-36d9-4a9c-a685-e0d7df3c399a" width="350">|


## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

adhiamboperes committed:
Fix forward databinding error from upstream

adhiamboperes committed:
Add forward navigation

adhiamboperes committed:
Remove modules from upstream

adhiamboperes committed:
Add navigation tests

adhiamboperes committed:
Fix alignment on tablet landscape

adhiamboperes committed:
Remove PrimeTopicAssetsControllerModule

adhiamboperes committed:
Move files to onboarding package

adhiamboperes committed:
Merge remote-tracking branch 'upstream/onboarding-learner-intro-screen' into onboarding-audio-language-screen

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileActivity.kt
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileActivityPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragment.kt
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/IntroActivity.kt
#	app/src/main/java/org/oppia/android/app/onboarding/IntroActivityPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/IntroFragment.kt
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeActivity.kt
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeActivityPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeFragment.kt
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeFragmentPresenter.kt
#	app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Refactor layouts based on upstream changes.

adhiamboperes committed:
Remove unused imports

adhiamboperes committed:
Create style for language dropdown

adhiamboperes committed:
Refactor tablet layout styling

adhiamboperes committed:
Refactor rename AudioLanguageFragmentPresenter to V1 suffix

adhiamboperes committed:
Add tests

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into new-onboarding-screen

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragmentPresenter.kt

adhiamboperes committed:
Merge branch 'new-onboarding-screen' into onboarding-profile-type-screen

adhiamboperes committed:
Merge remote-tracking branch 'upstream/onboarding-profile-type-screen' into create-profile-screen

adhiamboperes committed:
Fix merge conflict

adhiamboperes committed:
Merge branch 'new-onboarding-screen' into onboarding-profile-type-screen

adhiamboperes committed:
Merge branch 'onboarding-profile-type-screen' into create-profile-screen

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

adhiamboperes committed:
Create new style for error messages

adhiamboperes committed:
Add textwatcher test

adhiamboperes committed:
Fix small issues

adhiamboperes committed:
Fix small issues

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

# Conflicts:
#	app/src/main/res/values/styles.xml
#	app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt

adhiamboperes committed:
Fix merge conflicts

adhiamboperes committed:
Merge remote-tracking branch 'upstream/onboarding-learner-intro-screen' into onboarding-audio-language-screen

adhiamboperes committed:
Merge branch 'onboarding-audio-language-screen' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/options/AudioLanguageFragment.kt
#	app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt

adhiamboperes committed:
Create sole learner profile with tests

adhiamboperes committed:
Create profile onboarding event logs

adhiamboperes committed:
Exit app for sole learner profile

adhiamboperes committed:
Add login pathway for sole learner

adhiamboperes committed:
fix import order

adhiamboperes committed:
Add tests for migrated login routes

adhiamboperes committed:
Add tests for migrated login routes

adhiamboperes committed:
Log onboarding started event

adhiamboperes committed:
Update test initialization for onboarding v2 off

adhiamboperes committed:
Update test initialization for onboarding v2 off

adhiamboperes committed:
Merge branch 'develop' into new-onboarding-screen

adhiamboperes committed:
Merge branch 'new-onboarding-screen' into onboarding-profile-type-screen

adhiamboperes committed:
Merge branch 'onboarding-profile-type-screen' into create-profile-screen

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

adhiamboperes committed:
Merge branch 'onboarding-learner-intro-screen' into onboarding-audio-language-screen

adhiamboperes committed:
Merge branch 'onboarding-audio-language-screen' into onboarding-profile-domain-config

# Conflicts:
#	model/src/main/proto/oppia_logger.proto
#	testing/src/main/java/org/oppia/android/testing/logging/EventLogSubject.kt
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes committed:
Replace Lifecycle.State check with activity.isFinishing

adhiamboperes committed:
Flatten layout

adhiamboperes committed:
Resolve merge conflicts

adhiamboperes committed:
Merge branch 'onboarding-profile-type-screen' into create-profile-screen

adhiamboperes committed:
Adjust profile picture prompt

adhiamboperes committed:
Add assertions for image loaded

adhiamboperes committed:
Replace Lifecycle.State check with activity.isFinishing

adhiamboperes committed:
Addressed reviewer comments

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt

adhiamboperes committed:
Merge branch 'onboarding-learner-intro-screen' into onboarding-audio-language-screen

adhiamboperes committed:
Fix tests

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Add missing test initialization

adhiamboperes committed:
Merge branch 'onboarding-audio-language-screen' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/IntroFragment.kt

adhiamboperes committed:
Fix merge conflict

adhiamboperes committed:
Add missing bazel dependency

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/home/HomeActivityTest.kt:
```diff
@@ -1892,6 +1967,77 @@ class HomeActivityTest {
+      // no profile chooser.
+      // Using Lifecycle.State.DESTROYED instead of activity.isFinishing because the app is already
+      // closed, therefore we cannot run `onActivity`
+      assertThat(scenario.state).isEqualTo(Lifecycle.State.DESTROYED)
```
Replace with isFinishing

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -192,6 +200,11 @@ class ProfileManagementController @Inject constructor(
     return profileDataStore.transformAsync(GET_PROFILE_PROVIDER_ID) {
       val profile = it.profilesMap[profileId.internalId]
       if (profile != null) {
+        if (enableOnboardingFlowV2.value) {
+          if (profile.profileType.equals(ProfileType.PROFILE_TYPE_UNSPECIFIED)) {
+            updateProfileType(profileId, computeProfileType(profile.isAdmin, profile.pin))
+          }
+        }
```
Rewrite more elegantly

adhiamboperes commented on app/src/main/java/org/oppia/android/app/home/HomeActivity.kt:
```diff
@@ -106,4 +98,24 @@ class HomeActivity :
   }
+
+  override fun exitProfile(profileType: ProfileType) {
+    if (enableOnboardingFlowV2.value && profileType == ProfileType.SOLE_LEARNER) {
```
Condition should be checked before updating the isEnabled state of the OnBackPressedCallback: Ref https://stackoverflow.com/a/75462987/11396524

adhiamboperes committed:
Complete login migration route tests

adhiamboperes committed:
Fix tests/flows broken by changes

adhiamboperes committed:
Add tests for end profile onboarding event log

adhiamboperes committed:
Add tests for profile creation errors

adhiamboperes committed:
General cleanup

adhiamboperes committed:
General cleanup

adhiamboperes committed:
Revert breaking change

adhiamboperes committed:
Add kdoc

adhiamboperes committed:
Fix static check failures

adhiamboperes committed:
Fix failing event log events

adhiamboperes committed:
Fix failing app deprecation tests

adhiamboperes committed:
Refactor the app startup state

adhiamboperes committed:
Fix event logs

adhiamboperes committed:
Revert changes to app init

adhiamboperes committed:
Enforce v2 onboarding flow

adhiamboperes committed:
Enforce conditional profile exit for sole learner

adhiamboperes committed:
Revert unnecessary test file changes

adhiamboperes committed:
Fix proto field case

adhiamboperes commented on app/src/main/java/org/oppia/android/app/onboarding/AudioLanguageFragmentPresenter.kt:
```diff
@@ -45,6 +46,14 @@ class AudioLanguageFragmentPresenter @Inject constructor(
+      fragment.startActivity(intent)
+      fragment.activity?.finish()
+    }
+
```
Maybe login to profile here

adhiamboperes commented on app/src/main/java/org/oppia/android/app/onboarding/IntroActivityPresenter.kt:
```diff
@@ -10,9 +10,12 @@ import javax.inject.Inject
 
+/** Argument key for bundling the profileId. */
+const val PROFILE_ID_ARGUMENT_KEY = "internal_profile_id"
+
```
Move to proto

adhiamboperes commented on app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeFragmentPresenter.kt:
```diff
@@ -34,6 +34,7 @@ class OnboardingProfileTypeFragmentPresenter @Inject constructor(
       profileTypeSupervisorNavigationCard.setOnClickListener {
         val intent = ProfileChooserActivity.createProfileChooserActivity(activity)
         fragment.startActivity(intent)
+        activity.finishAffinity()
```
Revert

adhiamboperes commented:
@BenHenning, PTAL. The CI checks seem stuck on this PR

adhiamboperes committed:
Fix SplashActivityTests

adhiamboperes committed:
Merge branch 'develop' into onboarding-profile-type-screen

# Conflicts:
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Update test_file_exemptions to the new format

adhiamboperes committed:
Fix indent

adhiamboperes committed:
Merge branch 'onboarding-profile-type-screen' into create-profile-screen

# Conflicts:
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Update test_file_exemptions

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

# Conflicts:
#	model/src/main/proto/arguments.proto
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Address reviewer comment

adhiamboperes committed:
Merge branch 'onboarding-learner-intro-screen' into onboarding-audio-language-screen

# Conflicts:
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Address reviewer comments

adhiamboperes committed:
Populate the language dropdown list

adhiamboperes committed:
Merge branch 'onboarding-audio-language-screen' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/home/HomeActivity.kt
#	app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/AudioLanguageFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/testing/NavigationDrawerTestActivity.kt
#	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Fix test file exemption path

adhiamboperes committed:
Add missing bazel dep

BenHenning commented:
Hmm @adhiamboperes looking at the Bazel tests (which have better timeout cancellation than Gradle), I see at least ``SplashActivityTest`` is timing out at 5 minutes. I know that historically it takes much less time than that to run--have you tried running it locally? Changing anything with early app navigation is a bit tricky because there are a few things that need to be absolutely ordered perfectly for downstream activities to initialize properly. CI might be catching real failing tests; I'd focus on checking each of the failing Bazel tests locally to make sure they're actually passing as expected before delving too far into the Gradle tests.

I did re-run the Gradle workflows (and couldn't find anything about recent GitHub Actions issues), but if the timeout issues are real then they'll probably run into the same problem again.

adhiamboperes committed:
Add dropdown view id

adhiamboperes committed:
Add dropdown view id

adhiamboperes committed:
Hook up app language options

adhiamboperes committed:
Hook up app language options

adhiamboperes committed:
Merge branch 'develop' into onboarding-profile-type-screen

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/android/app/activity/ActivityComponentImpl.kt
#	app/src/main/java/org/oppia/android/app/fragment/FragmentComponentImpl.kt
#	app/src/main/res/values-night/color_palette.xml
#	app/src/main/res/values/color_defs.xml
#	app/src/main/res/values/color_palette.xml
#	app/src/main/res/values/component_colors.xml
#	model/src/main/proto/screens.proto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes committed:
Replace png images with svgs

adhiamboperes committed:
Add exemptions for new svgs

adhiamboperes committed:
Pull upstream and fix conflicts

adhiamboperes committed:
Change default avatar bg in v2

adhiamboperes committed:
Fix profile prompt background

adhiamboperes committed:
Merge branch 'develop' into create-profile-screen

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/android/app/activity/ActivityComponentImpl.kt
#	app/src/main/java/org/oppia/android/app/fragment/FragmentComponentImpl.kt
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeFragmentPresenter.kt
#	app/src/main/res/values-night/color_palette.xml
#	app/src/main/res/values/color_palette.xml
#	app/src/main/res/values/component_colors.xml
#	app/src/main/res/values/strings.xml
#	app/src/main/res/values/styles.xml
#	app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingProfileTypeFragmentTest.kt
#	model/src/main/proto/screens.proto
#	scripts/assets/test_file_exemptions.textproto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes committed:
Fix deprecated api

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

# Conflicts:
#	app/src/main/res/values/color_defs.xml
#	app/src/main/res/values/color_palette.xml
#	model/src/main/proto/screens.proto

adhiamboperes committed:
Fix kdoc

adhiamboperes committed:
Merge branch 'create-profile-screen' into onboarding-learner-intro-screen

adhiamboperes committed:
Merge remote-tracking branch 'upstream/onboarding-learner-intro-screen' into onboarding-audio-language-screen

adhiamboperes committed:
Fix test_file_exemption

adhiamboperes committed:
Merge branch 'develop' into onboarding-audio-language-screen

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/IntroFragmentPresenter.kt
#	app/src/main/res/values/styles.xml
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Replace ConstraintLayout with FrameLayout

adhiamboperes committed:
Create language dropdown list and the default selection

adhiamboperes committed:
Make otter graphic respond to rtl layouts

adhiamboperes committed:
Add mechanisms to check if a default profile exists: wip

adhiamboperes committed:
Pass the created profileId  and update profile details

Pass the OnboardingFragmentPresenter through intents and bundles

Added a new function to update multiple profile details at once.

adhiamboperes committed:
Refactored IntroFragment arguments to proto, and decorated with the profileId

Passing the profileId to AudioLanguageActivity eliminates the need to fetch it, when required for audio language operations.

adhiamboperes committed:
Prime AudioLanguageActivity to receive a profileId args

Also, remove instances of non-null assertion on internal profileId

adhiamboperes committed:
Fix merge conflicts

adhiamboperes committed:
Refactor to the new central hasProtoExtra

adhiamboperes committed:
Remove redundant non-null assertion

adhiamboperes committed:
Fix duplicating default profile creation

adhiamboperes committed:
Bind audiolanguage pre-selection to the dropdown

adhiamboperes committed:
Fix persist selection on config change

adhiamboperes committed:
Setup audio language pre-selection logic

adhiamboperes committed:
Fix lint errors

adhiamboperes committed:
Cleanup language selection and profile creation

adhiamboperes committed:
Fix missing kdocs

adhiamboperes committed:
Update profile error message view

Use an observable field to dynamically update errors.

adhiamboperes committed:
Refactor IntroActivity intent creation

Pass extras creation to the calling method

adhiamboperes committed:
Update intent tests to reflect new intent extras

adhiamboperes committed:
Add tests for language switching

adhiamboperes committed:
Add helper for creating a default empty profile

adhiamboperes committed:
Update updateNewProfileDetails() kdocs

adhiamboperes committed:
Helper for creating the default empty profile in tests

adhiamboperes committed:
Tests for the new profile update api

adhiamboperes committed:
Additional tests for the profile creation flow

adhiamboperes committed:
Add tests for language selection

Tests for locale, config change, selections and default selection.

adhiamboperes committed:
Fix missing kdoc

adhiamboperes committed:
Fix self-review pass issues.

adhiamboperes committed:
Exempt OnboardingFragmentTest from locale checks

adhiamboperes committed:
Fix failing tests

adhiamboperes commented on model/src/main/proto/profile.proto:
```diff
@@ -87,6 +87,12 @@ message Profile {
+  ProfileType profile_type = 19;
+
+  // Indicates whether this user has completed the onboarding flow.
+  bool already_onboarded_profile = 20;
```
Should be started and ended.

adhiamboperes committed:
Revert conflicting/already implemented UI layer code

adhiamboperes committed:
Temporary comments

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/IntroActivity.kt
#	app/src/main/java/org/oppia/android/app/onboarding/IntroActivityPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/IntroFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/options/AudioLanguageActivity.kt
#	app/src/main/java/org/oppia/android/app/options/AudioLanguageActivityPresenter.kt
#	app/src/main/java/org/oppia/android/app/options/AudioLanguageFragment.kt
#	app/src/main/java/org/oppia/android/app/translation/AppLanguageResourceHandler.kt
#	app/src/sharedTest/java/org/oppia/android/app/onboarding/IntroFragmentTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/options/OptionsFragmentTest.kt
#	domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt
#	domain/src/test/java/org/oppia/android/domain/profile/ProfileManagementControllerTest.kt
#	model/src/main/proto/oppia_logger.proto
#	model/src/main/proto/profile.proto
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Fix leftover refactor issue

adhiamboperes committed:
Complete the sole learner onboarding flow

Make sure all expected events are logged, plus tests.

adhiamboperes committed:
Update admin profile profileType in v2

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt
#	domain/src/test/java/org/oppia/android/domain/profile/ProfileManagementControllerTest.kt

adhiamboperes committed:
Destroy onboarding activities after onboarding completed.

adhiamboperes committed:
Fix failing SplashActivityTests

adhiamboperes committed:
Fix failing SplashActivityTests + finish migration pathways

adhiamboperes committed:
Onboard existing learner profiles

adhiamboperes committed:
Toggle homescreen with multiple classrooms flag

adhiamboperes committed:
General cleanup

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/home/HomeActivity.kt
#	app/src/sharedTest/java/org/oppia/android/app/onboarding/IntroFragmentTest.kt

adhiamboperes committed:
General cleanup

adhiamboperes committed:
Fix failing test

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

BenHenning commented:
Will need to take a pass on this a bit later since #5457 had a bunch of comments. Please make sure to address anything from #5457 that's relevant here, and then assign back.

adhiamboperes committed:
Address part of reviewer comments

adhiamboperes committed:
Fix lint errors

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/AudioLanguageFragmentPresenter.kt

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' of github.com:oppia/oppia-android into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

adhiamboperes committed:
Create test suite for TextInputLayoutBindingAdapters

adhiamboperes committed:
Create test suite for TextInputLayoutBindingAdapters

adhiamboperes committed:
Address more reviewer comments.

adhiamboperes committed:
Address more reviewer comments.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-language-domain-config

adhiamboperes committed:
Fixed static check failures

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

adhiamboperes committed:
Fix empty strings assertion

adhiamboperes committed:
Addressed test comments

adhiamboperes committed:
Use proto for app language

adhiamboperes committed:
Refactor language binding logic

adhiamboperes committed:
Allow user to progress on continue clicked.

adhiamboperes committed:
Fix failing test

adhiamboperes committed:
Fix failing audio language tests

adhiamboperes committed:
Fix failing profile creation tests

adhiamboperes committed:
Temporarily remove failing adapter tests

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/AudioLanguageFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileActivityPresenter.kt

adhiamboperes commented:
Resolved conflicts from upstream. Waiting for CI to pass.

adhiamboperes reviewed (commented):
Self review comments

adhiamboperes commented:
@BenHenning, assigning this back.

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

# Conflicts:
#	app/src/main/java/org/oppia/android/app/player/audio/AudioFragmentPresenter.kt

adhiamboperes committed:
Update profile creation exceptions and tests

adhiamboperes committed:
Properly set up binding adapter tests

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt
#	scripts/assets/test_file_exemptions.textproto

adhiamboperes committed:
Fix self review issues

adhiamboperes committed:
Remove unused code

adhiamboperes committed:
Fix nit

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

adhiamboperes committed:
Remove duplicate exemption

adhiamboperes committed:
Remove leftover profile audiolanguage

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

adhiamboperes committed:
Integrate classrooms and onboarding

adhiamboperes committed:
Fix Admin profile creation and onboarding

adhiamboperes committed:
Add missing argument proto

adhiamboperes committed:
Update broken tests

adhiamboperes committed:
Fix profile migration bugs

adhiamboperes committed:
Fix failing audio language tests

BenHenning commented on testing/src/main/java/org/oppia/android/testing/logging/EventLogSubject.kt:
```diff
@@ -2400,6 +2455,36 @@ class EventLogSubject private constructor(
+     * This method never fails since the underlying property defaults to empty string if it's not
+     * defined in the context.
+     */
+    fun hasProfileIdThat(): Subject = assertThat(actual.profileId)
```
Should this maybe return ``LiteProtoSubject`` instead?

adhiamboperes replied:
Yes, updated

BenHenning commented on testing/src/test/java/org/oppia/android/testing/profile/ProfileTestHelperTest.kt:
```diff
@@ -138,6 +139,47 @@ class ProfileTestHelperTest {
+  }
+
+  @Test
+  fun testUpdateProfile_updateProfileType_checkIsSuccessful() {
```
```suggestion
  fun testUpdateProfile_updateProfileType_profileTypeShouldBeUpdated() {
```

This should be more specific as to what is succeeding.

Please update all other new test names in this PR to match this level of specificity.

adhiamboperes replied:
Updated.

BenHenning commented on testing/src/test/java/org/oppia/android/testing/profile/ProfileTestHelperTest.kt:
```diff
@@ -138,6 +139,47 @@ class ProfileTestHelperTest {
+  }
+
+  @Test
+  fun testProfileOnboarding_markOnboardingCompleted_checkIsSuccessful() {
```
For this & the above: can they instead verify that the event was successfully logged rather than the provider succeeds (or better yet, have tests for both)? I think verifying both the provider result and outcome is important for consistency and correctness.

adhiamboperes replied:
Done

BenHenning commented on model/src/main/proto/profile.proto:
```diff
@@ -93,6 +93,12 @@ message Profile {
+  // Indicates whether this profile has started the onboarding flow.
+  bool started_profile_oboarding = 21;
+
+  // Indicates whether this profile has completed the onboarding flow.
```
What does started and completed mean? This is a bit difficult to grok for me because where users start and finish the onboarding flows depends on profile type, so the current framing is ambiguous and needs more specificity, I think.

adhiamboperes replied:
started - Indicates that the profile has viewed the relevant onboarding introduction screen.

completed - Indicates that the profile has reached the home screen for the first time.

I have updated the proto field comments.

BenHenning commented on model/src/main/proto/arguments.proto:
```diff
@@ -257,6 +260,9 @@ message ReadingTextSizeFragmentStateBundle {
   AudioLanguage audio_language = 1;
+
+  // The internal Id associated with the profile.
+  int32 profile_id = 2;
```
Here & for the other new fields, prefer using ``ProfileId`` to avoid adding more work to #4865.

adhiamboperes replied:
Cleaned these up in favor of the decorator as per #4865.

BenHenning commented on app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt:
```diff
@@ -149,17 +160,23 @@ class ClassroomListFragmentPresenter @Inject constructor(
+    if (enableOnboardingFlowV2.value) {
+      subscribeToProfileResult(profileId)
+    } else {
+      logAppOnboardedEvent(profileId)
```
We want to log this for all cases where the user successfully finishes the onboarding flow (there's a bug with it currently whereby it logs for every home screen, and we're looking into it).

It acts as a top-level app health metric, so please make sure to account for it with the new onboarding changes.

adhiamboperes replied:
`subscribeToProfileResult()` > `processProfileResult()` > `handleProfileOnboardingState(profile: Profile)`  ensures that all users who complete the new onboarding flow log a `started_profile_onboarding` and `completed_profile_onboarding` event.  At the same time, the first profile on the app(supervisor or sole learner) will additionally log the app onboarding complete event(existing event).

In this PR, I have retained the legacy onboarding flow to only log the app onboarding event.

I implemented this per TDD, although it is different from your comment(sounds like we want all profiles to log all 3 events?). Could you please clarify the intended behaviour?

adhiamboperes replied:
Seperately, I experienced the bug you mention, and given some context I may be able to pick up debugging from my prior attempt at solving the https://github.com/oppia/oppia-android/issues/5434 (closed due to reverting adjacent changes that may have triggered the bug for me).

adhiamboperes replied:
As per the discussion in the meeting, I left the implementation of the logAppOnboardedEvent as is.

BenHenning replied:
Ah, yes I think keeping the existing behavior is reasonable since fixing an existing event is outside the scope of this PR.

BenHenning commented on app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt:
```diff
@@ -297,12 +314,71 @@ class ClassroomListFragmentPresenter @Inject constructor(
+        logAppOnboardedEvent(profileId)
+      }
+    }
+  }
```
I'm quite uncomfortable, in general, with logging events in the app layer because its lifecycle is incredibly complex. It's very easily to accidentally duplicate logs (e.g. due to activity recreation, etc.). Can we log this in the domain layer, instead?

adhiamboperes replied:
I have moved this to the controller.

BenHenning commented on app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt:
```diff
@@ -297,12 +314,71 @@ class ClassroomListFragmentPresenter @Inject constructor(
   }
+
+  private fun handleBackPress(profileType: ProfileType) {
+    activity.onBackPressedDispatcher.addCallback(
```
If the profile provider changes multiple times, it will lead to multiple callbacks being registered. Does this code correctly handle that scenario?

adhiamboperes replied:
Refactored function to remove this callback from any OnBackPressedDispatcher it is currently added to after every call.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/ExitProfileListener.kt:
```diff
@@ -0,0 +1,13 @@
+package org.oppia.android.app.home
+
+import org.oppia.android.app.model.ProfileType
+/** Listener for when a user wishes to exit their profile. */
```
Nit: empty line above this please.

adhiamboperes replied:
Fixed

BenHenning commented on app/src/main/java/org/oppia/android/app/home/ExitProfileListener.kt:
```diff
@@ -0,0 +1,13 @@
+  /**
+   * Called when back press is clicked on the HomeScreen.
+   *
+   * A SOLE_LEARNER exits the app completely while other [ProfileType]s are routed to the
+   * [ProfileChooserActivity].
```
Suggest instead saying that routing behavior may change based on the profile type rather than specifying the actual implementation behavior (which seems a bit too specific at the interface level).

adhiamboperes replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt:
```diff
@@ -142,6 +155,52 @@ class HomeFragmentPresenter @Inject constructor(
+        profile.profileType == ProfileType.SUPERVISOR
+      ) {
+        appStartupStateController.markOnboardingFlowCompleted()
+        logAppOnboardedEvent(profileId)
```
Ditto here--we have multiple places where events need to be logged correctly in the UI layer. It seems *really* hard to verify this is actually correct--logging at the domain layer will likely be much easier to validate as correct and keep correct.

adhiamboperes replied:
Moved profile onboarding event to the controller and kept the implementation of the logAppOnboardedEvent.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt:
```diff
@@ -209,4 +268,15 @@ class HomeFragmentPresenter @Inject constructor(
   }
+
+  private fun handleBackPress(profileType: ProfileType) {
+    activity.onBackPressedDispatcher.addCallback(
```
Ditto here--callback can be added multiple times.

adhiamboperes replied:
Refactored the code handling profile update and the callback.

BenHenning commented on app/src/main/java/org/oppia/android/app/onboarding/AudioLanguageFragmentPresenter.kt:
```diff
@@ -158,6 +159,30 @@ class AudioLanguageFragmentPresenter @Inject constructor(
       }
   }
 
+  private fun loginToProfile(profileId: ProfileId) {
```
```suggestion
  private fun logInToProfile(profileId: ProfileId) {
```

Nit: should be verb form.

adhiamboperes replied:
Refactored as suggested.

BenHenning commented on app/src/main/java/org/oppia/android/app/onboarding/IntroFragmentPresenter.kt:
```diff
@@ -67,4 +75,13 @@ class IntroFragmentPresenter @Inject constructor(
   }
+
+  private fun markProfileOnboardingStarted(profileId: ProfileId) {
+    profileManagementController.markProfileOnboardingStarted(profileId)
```
Ditto for this event: prefer logging in the domain layer, or via a UI call to ensure the user actually saw the correct flow prior to logging the event (though this call needs to be idempotent to be resilient against multiple calls which is possible in the UI layer).

adhiamboperes replied:
Fixed

BenHenning commented on app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragmentPresenter.kt:
```diff
@@ -252,8 +252,7 @@ class OnboardingFragmentPresenter @Inject constructor(
 
   private fun createDefaultProfile() {
     profileManagementController.addProfile(
-      name = "Admin", // TODO(#4938): Refactor to empty name once proper admin profile creation flow
```
Does  this need to be contingent based on whether the new onboarding flow flag is enabled?

adhiamboperes replied:
The onboarding flow version has been gated at fragment level, such that we have presenter v1, and this, v2. Please see: https://github.com/oppia/oppia-android/blob/4730edb0263135e88dab08ab0f6dc724a1585a67/app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragment.kt#L36-L40

BenHenning commented on app/src/main/java/org/oppia/android/app/onboarding/OnboardingProfileTypeFragmentPresenter.kt:
```diff
@@ -66,4 +89,13 @@ class OnboardingProfileTypeFragmentPresenter @Inject constructor(
+  private fun markProfileOnboardingStarted(profileId: ProfileId) {
+    profileManagementController.markProfileOnboardingStarted(profileId)
+
+    analyticsController.logLowPriorityEvent(
```
Ditto for this event.

adhiamboperes replied:
Refactored

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileChooserActivityPresenter.kt:
```diff
@@ -3,27 +3,45 @@ package org.oppia.android.app.profile
+        isAdmin = true
+      )
+    } else {
+      // TODO(#482): Ensures that an admin profile is present. Remove when there is proper admin account creation.
```
I think we can reframe this as: 'Remove when the new onboarding flow can be finalized as that properly creates an admin profile.' Your work *is* the solution desired by this comment.

adhiamboperes replied:
Refactored

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileChooserFragmentPresenter.kt:
```diff
@@ -267,4 +254,54 @@ class ProfileChooserFragmentPresenter @Inject constructor(
+    activity.startActivity(intent)
+  }
+
+  private fun loginToProfile(profile: Profile) {
```
Nit: ``logIn`` here, as well.

adhiamboperes replied:
Updated.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! I reviewed a bit over half of the PR. Will take a broader pass on the next round since there were a bunch of comments that would be nice to address, first.

Edit: Also, it seems that CI is failing--PTAL.

adhiamboperes committed:
Add test for arabic inTextInputLayoutBindingAdaptersTest

adhiamboperes committed:
Refactor formatting

adhiamboperes committed:
Address reviewer comment

adhiamboperes committed:
Cleanup language selection impl

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-language-domain-config

adhiamboperes committed:
Address general comments

adhiamboperes committed:
Nit

adhiamboperes committed:
Nit

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Merge branch 'onboarding-profile-domain-config' of github.com:oppia/oppia-android into onboarding-profile-domain-config

adhiamboperes committed:
remove unused import

adhiamboperes committed:
Refactor event log management

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-language-domain-config

adhiamboperes committed:
Refactor function

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'develop' into onboarding-language-domain-config

adhiamboperes committed:
Refactor language selection to use supported languages only

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' of github.com:oppia/oppia-android into onboarding-language-domain-config

adhiamboperes committed:
Merge branch 'onboarding-language-domain-config' into onboarding-profile-domain-config

adhiamboperes committed:
Refactor test to use existing eventlog util function

adhiamboperes committed:
Null check and readability improvements

adhiamboperes committed:
Refactor test setup

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-profile-domain-config

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Fix failing profile chooser tests

adhiamboperes committed:
Resolve merge conflicts

adhiamboperes committed:
Refactor backpress callback

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-profile-domain-config

adhiamboperes committed:
Fix synchronisation issues caused by createEmptyComposeRule()

adhiamboperes committed:
Revert formatting changes

adhiamboperes committed:
Refactor callback code

adhiamboperes committed:
Revert formatting only changes

adhiamboperes committed:
Add missing tests

adhiamboperes committed:
Pull latest changes

adhiamboperes commented:
@BenHenning, PTAL.

adhiamboperes committed:
Fix merge issues

BenHenning commented on app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt:
```diff
@@ -265,36 +279,142 @@ class SplashActivityPresenter @Inject constructor(
+    activity.startActivity(intent)
+  }
+
+  private fun loginToProfile(profileId: ProfileId) {
```
```suggestion
  private fun logInToProfile(profileId: ProfileId) {
```

Nit for consistency & correct verb tense. Ditto elsewhere in the PR.

adhiamboperes replied:
Updated -- other occurances previously updated in profileChooser and AudioLanguageFragment.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt:
```diff
@@ -1049,6 +1058,72 @@ class SplashActivityTest {
+  }
+
+  @Test
+  @RunOn(TestPlatform.ESPRESSO)
```
Why can't this run on Robolectric?

adhiamboperes replied:
There seems to be a synchronizing issue with this test. I have updated the test code itself:
```kotlin
 @Test
  @RunOn(TestPlatform.ESPRESSO)
  fun testSplashActivity_onboardingV2Enabled_onboardedSoleLearnerProfile_routesToHomeActivity() {
    simulateAppAlreadyOnboarded()
    initializeTestApplication(onboardingV2Enabled = true)
    profileTestHelper.addOnlyAdminProfileWithoutPin()
    testCoroutineDispatchers.runCurrent()

    val profileId = ProfileId.newBuilder().setInternalId(0).build()
   monitorFactory.ensureDataProviderExecutes(
      profileTestHelper.updateProfileType(profileId, ProfileType.SOLE_LEARNER)
    )
    monitorFactory.ensureDataProviderExecutes(
      profileTestHelper.markProfileOnboardingStarted(profileId)
    )
    monitorFactory.ensureDataProviderExecutes(
      profileTestHelper.markProfileOnboardingEnded(profileId)
    )
    testCoroutineDispatchers.advanceUntilIdle()

    launchSplashActivityPartially{
      intended(hasComponent(HomeActivity::class.java.name))
    }
  }
```

`profileTestHelper.markProfileOnboardingEnded(profileId)` never executes, causing the test to run indefinitely due to never entering an idle state. 
On Espresso, this passes but ocassionally fails with:

```
Wanted but not invoked:
observer.onChanged(<Capturing argument>);
-> at org.oppia.android.testing.data.DataProviderTestMonitor.ensureNextResultIsPresent(DataProviderTestMonitor.kt:117)
Actually, there were zero interactions with this mock.

at org.oppia.android.testing.data.DataProviderTestMonitor.ensureNextResultIsPresent(DataProviderTestMonitor.kt:117)
at org.oppia.android.testing.data.DataProviderTestMonitor.waitForNextResult(DataProviderTestMonitor.kt:59)
```

For any random DataProvider in the test.

adhiamboperes replied:
This turned out to be a dataprovider update loop, and I have fixed it.

BenHenning replied:
I am actually very interested in what the specific problems were, and if this is something we can try to document on the data providers wiki page. Let's discuss further.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -209,6 +218,11 @@ class ProfileManagementController @Inject constructor(
     return profileDataStore.transformAsync(GET_PROFILE_PROVIDER_ID) {
       val profile = it.profilesMap[profileId.internalId]
       if (profile != null) {
+        if (enableOnboardingFlowV2.value) {
```
Are there tests verifying this migration pathway and do they cover all potential scenarios? It seems they do, but I wanted to double check since it's essential that we get this part right (given it will affect all existing users of the app).

adhiamboperes replied:
You're correct, these scenarios were implicitly tested, but I have added more explicit tests for them.

BenHenning reviewed (commented):
Thanks @adhiamboperes! Just had a few small comments left, otherwise the PR LGTM.

adhiamboperes committed:
Resolve merge conflicts

adhiamboperes committed:
Address reviewer comments

adhiamboperes committed:
Run test on CI without espresso annotation

adhiamboperes committed:
Fix failing tests and livedata bug

adhiamboperes committed:
Add tests for profile type migration

BenHenning commented on testing/src/main/java/org/oppia/android/testing/profile/ProfileTestHelper.kt:
```diff
@@ -70,14 +70,13 @@ class ProfileTestHelper @Inject constructor(
    *
    * @returns the [AsyncResult] designating the result of attempting to log into the admin profile
```
It seems the KDoc needs to be updated with the recent changes.

adhiamboperes replied:
This is done.

BenHenning reviewed (approved):
Thanks @adhiamboperes! There just seems to be one small comment left. Feel free to address and resolve it on your own before submitting. I can take another pass if you need me to, but I don't expect that to be necessary.

BenHenning commented:
It looks like ``SplashActivityTest`` is timing out on Bazel--maybe on Gradle, too (hard to tell with the results since the Gradle tests are behaving oddly). I'm going ahead and updating the branch so that we can get fresh results.

BenHenning committed:
Merge branch 'develop' into onboarding-profile-domain-config

adhiamboperes committed:
Add tests for complete  app onboarding

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-profile-domain-config

adhiamboperes committed:
Added test coverage for onboarding events

adhiamboperes committed:
Merge branch 'onboarding-profile-domain-config' of github.com:oppia/oppia-android into onboarding-profile-domain-config

adhiamboperes commented:
@BenHenning, thanks for reviewing! I have adressed the remaining comment, and also added the comment on the unobserved dataprovider as discussed in the meeting. The tests pass, and I am proceeding with merging the PR.
